scripts/abins/input/jsonloader.py

```diff
@@ -0,0 +1,145 @@
+# Mantid Repository : https://github.com/mantidproject/mantid
+#
+# Copyright &copy; 2024 ISIS Rutherford Appleton Laboratory UKRI,
+#   NScD Oak Ridge National Laboratory, European Spallation Source,
+#   Institut Laue - Langevin & CSNS, Institute of High Energy Physics, CAS
+# SPDX - License - Identifier: GPL - 3.0 +
+
+from enum import auto, Enum
+import json
+from pathlib import Path
+from typing import Dict
+
+from euphonic import QpointPhononModes
+import numpy as np
+
+from .abinitioloader import AbInitioLoader
+from .euphonicloader import EuphonicLoader
+from abins.abinsdata import AbinsData
+from abins.constants import COMPLEX_TYPE, FLOAT_TYPE
+from abins.parameters import sampling as sampling_parameters
+from dos.load_euphonic import euphonic_calculate_modes
+
+# json-stream implementation converts data lazily so we can quickly check the
+# top-level keys even if a data file is huge.
+try:
+    from json_stream import load as json_load
+except ImportError:
+    from json import load as json_load
+
+
+class PhononJSON(Enum):
+    EUPHONIC_FREQUENCIES = auto()
+    EUPHONIC_MODES = auto()
+    EUPHONIC_FORCE_CONSTANTS = auto()
+    ABINS_DATA = auto()
+    UNKNOWN = auto()
+
+
+abins_supported_json_formats = {PhononJSON.EUPHONIC_MODES, PhononJSON.EUPHONIC_FORCE_CONSTANTS, PhononJSON.ABINS_DATA}
+
+
+class JSONLoader(AbInitioLoader):
+    """Get frequencies/eigenvalues from a JSON file using Euphonic"""
+
+    @property
+    def _ab_initio_program(self) -> str:
+        return "JSON"
+
+    @staticmethod
+    def check_json_format(json_file: str | Path) -> PhononJSON:
+        """Check if JSON file is a known phonon data format"""
+
+        for class_key in "__euphonic_class__", "__abins_class__":
+            with open(json_file, "r") as fd:
+                data = json_load(fd)
+                data_class = data.get(class_key)
+
+            if data_class is not None:
+                break
+        else:
+            class_key = ""
+
+        match class_key, data_class:
```

Comment:
Interesting, never seen this before! Was it introduced in a recent python version?